.github/workflows/industrial_ci_action.yml

```
@@ -10,19 +10,50 @@ jobs:
     strategy:
       matrix:
         env:
-          - {ROS_DISTRO: foxy, ROS_REPO: main}
-          - {ROS_DISTRO: foxy, ROS_REPO: testing}
+          - {
+            CACHE_PREFIX: 'foxy-main-debug-ccov',
+            ROS_DISTRO: foxy, ROS_REPO: main, CCOV_UPLOAD: true,
+            CMAKE_ARGS: "-DCMAKE_BUILD_TYPE=Debug -DCMAKE_C_FLAGS='--coverage' -DCMAKE_CXX_FLAGS='--coverage'",
+            AFTER_RUN_TARGET_TEST: './.ci.prepare_codecov',
+            ADDITIONAL_DEBS: 'lcov'
+          }
+          - {
+            CACHE_PREFIX: 'foxy-testing', ROS_DISTRO: foxy, ROS_REPO: testing, 
+            CCOV_UPLOAD: false
+          }
     env:
       UPSTREAM_WORKSPACE: warehouse_ros_mongo.repos
-      CCACHE_DIR: ~/.ccache
+      AFTER_SETUP_UPSTREAM_WORKSPACE: 'vcs pull $BASEDIR/upstream_ws/src'
+      CCACHE_DIR: "${{ github.workspace }}/.ccache"
+      BASEDIR: ${{ github.workspace }}/.work
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v2
-      - uses: actions/cache@v2
+      - name: cache upstream_ws
+        uses: pat-s/always-upload-cache@v2.1.3
+        with:
+          path: ${{ env.BASEDIR }}/upstream_ws
+          key: upstream_ws-${{ matrix.env.CACHE_PREFIX }}-${{ hashFiles('warehouse_ros_mongo.repos') }}-${{ github.run_id }}
+          restore-keys: |
+            upstream_ws-${{ matrix.env.CACHE_PREFIX }}-${{ hashFiles('warehouse_ros_mongo.repos') }}
+      - name: cache ccache
+        uses: pat-s/always-upload-cache@v2.1.3
         with:
           path: ${{ env.CCACHE_DIR }}
-          key: ccache-${{ matrix.env.ROS_DISTRO }}-${{ matrix.env.ROS_REPO }}-${{ github.sha }}  
+          key: ccache-${{ matrix.env.CACHE_PREFIX }}-${{ github.sha }}  
```

Of course, these change then as well:
```suggestion
          key: ccache-${{ env.CACHE_PREFIX }}-${{ github.sha }}  
```